### ISSUE TITLE
It can break the web

### ISSUE THREAD

**Issue by zloirock** [mathiasbynens/String.prototype.at#6]
[`SugarJS`](https://sugarjs.com/docs/#/String/at) provides a method with the same name, but with another signature. Related https://github.com/tc39/proposal-flatMap/pull/56. BTW is this proposal still alive?

**Comment by michaelficarra** [comment]
It's only a problem if it's being conditionally defined. Is SugarJS checking for the existence of `String.prototype.at` before assigning/defining it?

**Comment by zloirock** [comment]
@michaelficarra why do you think that it's problem only in this case? You can use a method from SugarJS, but once after a patch release, your dependency can start use method with the same name from the standard library.

**Comment by michaelficarra** [comment]
@zloirock You don't seem to understand the issue. If I do

```js
String.prototype.at = function () { ... };
```

then introducing a native version does not break this code because my code overwrites it. Instead, I would have to do something like

```js
if (!String.prototype.at) {
  String.prototype.at = function () { ... };
}
```

*and* I would have to be defining it to behave differently than the native one.

**Comment by zloirock** [comment]
@michaelficarra I understood you, but I had some issues in `core-js` even in this case.

**Comment by ljharb** [comment]
So, to confirm, how does SugarJS define an "at" method?

**Comment by zloirock** [comment]
@ljharb seems unconditional replacement. But, as I wrote, I had many examples, mainly with ES6 methods, when something like that caused problems too.

**Comment by hax** [comment]
While the worst case is we need to rename the method, like `utf16CharAt` -- obviously it's not a great name but at least it's descriptive and unlikely to break the Web.

But first, is this proposal still alive?

**Comment by mathiasbynens** [comment]
@hax `utf16CharAt` already exists as `String.prototype.charAt`.

The whole point of this proposal is to have something that doesn’t deal with UTF-16 code units but rather with Unicode code points.

**Comment by hax** [comment]
@mathiasbynens Sorry, it's a really bad name. I try to use `utf16Char` to express a real unicode character encoded in UTF16 which may be two `char` in JS string 😓

Too hard to find a correct name...

It can't be `charAt` because it's already be taken and mean utf16 code unit.
It can't be `characterAt` because it's confused why `char` and `character` is different.
It can't be `utf32CharAt` because the result still a string encoded in UTF16.
It can't be `utf16CharAt` because the original `charAt` should use it.
...

Maybe `unicodeCharAt` ?

But it's still bad. I just realize we should use `char` in the name or it always confusing.

**Comment by mathiasbynens** [comment]
The idea is to provide an API that effectively hides the encoding that JS uses ([UCS-2/UTF-16](https://mathiasbynens.be/notes/javascript-encoding)) and instead allows the user to operate on Unicode code points, just like the string iterator does. `codePointAt`/`at` fail that test because the `position` arguments they accept are still UCS-2/UTF-16 indices.

But then again, maybe the string iterator is sufficient and we don’t need those APIs anymore.

**Comment by js-choi** [comment]
@mathiasbynens: For what it’s worth, string indexing by code points is still important to the [Web Annotation Data Model][], which chose to use code-point indexing on the recommendation of W3C Internationalization. A performant way to do this would still be very valuable both to our use case and probably also others’.

**Comment by Ohar** [comment]
I don't think that method's proposals should depends on which method are in some popular libraries. It is totally those libraries problem.
